### PR TITLE
PHP 7 Throwable Errors Exceptions

### PR DESCRIPTION
**Pull request by instabledesign**

### CODE REVIEW

**Comment by ElevenWilson** [comment]
40 words might be misspelled, please check them: AffichageException, amoutte, débogage, Encapsulé, encapsuler, etre, eval, executer, generatePassword, getMyObject, getPrevious, hierarchie, lang, l'encapsulation, l'object, logguer, LogicalException, LogicException, loging, maitrise, MyClass, MyObject, MyObjectException, obliguatoire, ParseError, parsez, passwordGenerator, personalisée, qualitée, responsabilitées, rethrow, RuntimeException, throwable, Throwable, traditionel, TypeError, typehint, UpdateContentException, url, UserFactoryException

**Comment by ElevenWilson** [comment]
31 words might be misspelled, please check them: amoutte, débogage, encapsuler, Encapsuler, eval, generatePassword, getMyObject, getPrevious, lang, l'encapsulation, logguer, LogicalException, LogicException, MyClass, MyObject, MyObjectException, ParseError, parsez, passwordGenerator, personalisée, qualitée, responsabilitées, rethrow, RuntimeException, throwable, Throwable, TypeError, typehint, UpdateContentException, url, UserFactoryException

**Comment by ElevenWilson** [comment]
31 words might be misspelled, please check them: amoutte, débogage, encapsuler, Encapsuler, eval, generatePassword, getMyObject, getPrevious, lang, l'encapsulation, logguer, LogicalException, LogicException, MyClass, MyObject, MyObjectException, ParseError, parsez, passwordGenerator, personalisée, qualitée, responsabilitées, rethrow, RuntimeException, throwable, Throwable, TypeError, typehint, UpdateContentException, url, UserFactoryException

**Comment by ElevenWilson** [comment]
29 words might be misspelled, please check them: amoutte, débogage, encapsuler, Encapsuler, eval, generatePassword, getMyObject, getPrevious, lang, l'encapsulation, logguer, LogicalException, LogicException, MyClass, MyObject, MyObjectException, ParseError, parsez, passwordGenerator, personalisée, rethrow, RuntimeException, throwable, Throwable, TypeError, typehint, UpdateContentException, url, UserFactoryException

File: _posts/fr/2018-02-11-throwable-error-exception.md [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+```php
+interface Throwable
+{
+    public function getMessage(): string;       // La raison de l'erreur
+    public function getCode(): int;             // Le code de l'erreur
+    public function getFile(): string;          // Le fichier dans lequel l'erreur à débuter
+    public function getLine(): int;             // La ligne à laquel l'erreur à débuter
+    public function getTrace(): array;          // Retourne la stack trace en array comme debug_backtrace()
+    public function getTraceAsString(): string; // Retourne la stack trace en chaine de caractère
+    public function getPrevious(): Throwable;   // Retourne le `Trowable` précédent
```
Comment (rmasclef) [inline _posts/fr/2018-02-11-throwable-error-exception.md]: typo -> t**h**rowable

File: _posts/fr/2018-02-11-throwable-error-exception.md [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+}
+```
+
+### Error
+
+`Error` est la classe de base de toutes les erreurs internes de PHP.
+
+Les plus communes sont :
+ - `ParseError`, qui est lancée quand on `require` ou `eval` un code qui contient une erreur de syntax.
+ - `TypeError`, qui est lancé quand le typehint d'un argument/retour d'une fonction n'est pas respectée. _Et également en `strict mode` quand on passe un nombre invalid d'arguments à une fonction native de PHP._
```
Comment (rmasclef) [inline _posts/fr/2018-02-11-throwable-error-exception.md]: lancée
respecté

File: _posts/fr/2018-02-11-throwable-error-exception.md [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+
+_Vous pourriez être amenés à throw des `Error` dans votre code si par exemple vous parsez un fichier et qu'il contient une erreur de syntaxe.
+Ou si vous avez une fonction avec un nombre de paramètres variable et que le nombre/type d'arguments n'est pas correct._
+
+### Exception
+
+`Exception` est la classe de base de toutes les exceptions utilisateurs.
+
+Il est très fréquent de lancer ou créer des `Exception`. 
+C'est d'ailleurs sur le fonctionnement et l'utilisation des `Exception` que nous nous concentrer.
```
Comment (rmasclef) [inline _posts/fr/2018-02-11-throwable-error-exception.md]: que nous allons nous concentrer ?

File: _posts/fr/2018-02-11-throwable-error-exception.md [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+
+Pour lancer une exception, il suffit d'utiliser le mot clé `throw`.
+
+```php
+throw new Exception('Mon message d\'erreur.');
+echo "Affichage d'un contenu texte.";
+```
+
+> Il faut savoir qu'une `Exception` interrompt l'exécution des instructions suivantes.
+> Dans l'exemple le `echo` ne seras pas exécuté.
```
Comment (rmasclef) [inline _posts/fr/2018-02-11-throwable-error-exception.md]: ne sera

File: _posts/fr/2018-02-11-throwable-error-exception.md [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+
+### Attraper une exception
+
+Pour attraper et gérer l'exception, il faut utiliser la structure `try` `catch`.
+```php
+try {
+    if (!$_GET['titre']) {
+        throw new Exception('Impossible d'afficher le titre. Le titre est requis.);
```
Comment (rmasclef) [inline _posts/fr/2018-02-11-throwable-error-exception.md]: `throw new \Exception('Impossible d\'afficher le titre. Le titre est requis.');`

Comment (rmasclef) [reply _posts/fr/2018-02-11-throwable-error-exception.md]: I'm not a fan of french text in code (I'm OK for french comments for the article purpose though)

File: _posts/fr/2018-02-11-throwable-error-exception.md [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+    echo '⚠ Une exception est survenue : ' . $e->getMessage();
+}
+``` 
+
+__⚠ Il est très important de bien choisir l'`Exception` que l'on veut lancer ou attraper, sinon la gestion d'erreur ne sera pas consistante.__
+
+**Également à savoir**
+
+La `LogicException` référence une erreur de code qui devrait, la plupart du temps, mener à un correctif sur le code.
+Attraper une `LogicalException` a généralement pour but d'afficher une page d'erreur et de logger en vue d'informer le développeur.
```
Comment (rmasclef) [inline _posts/fr/2018-02-11-throwable-error-exception.md]: LogicalException ? 🤔  LogicException

File: _posts/fr/2018-02-11-throwable-error-exception.md [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+
+La `RuntimeException` représente des erreurs durant l'exécution (donnée invalide, erreur d'une source de données).
```
Comment (rmasclef) [inline _posts/fr/2018-02-11-throwable-error-exception.md]: La ... représente **une erreur** survenue durant ...

File: _posts/fr/2018-02-11-throwable-error-exception.md [hunk 8]
```diff
@@ -0,0 +1,374 @@
...
+     * @return MyObject
+     */
+    public function getMyObject()
+    {
+        return $this->myObject;
+    }
+}
+```
+
+> Quand l'`MyObjectException` est attrapé, on peut récupérer l'objet `MyObject` via la méthode `getMyObject()`
```
Comment (rmasclef) [inline _posts/fr/2018-02-11-throwable-error-exception.md]: Quand `MyObjectException` est attrapée

File: _posts/fr/2018-02-11-throwable-error-exception.md [hunk 9]
```diff
@@ -0,0 +1,374 @@
...
+> ce qui permet de gérer encore plus précisément la gestion d'erreur.
```
Comment (rmasclef) [inline _posts/fr/2018-02-11-throwable-error-exception.md]: la tournure est bizarre, "gérer la gestion d'erreur" non ?

File: _posts/fr/2018-02-11-throwable-error-exception.md [hunk 10]
```diff
@@ -0,0 +1,374 @@
...
+    }
+}
+```
+
+> Quand l'`MyObjectException` est attrapé, on peut récupérer l'objet `MyObject` via la méthode `getMyObject()`
+> ce qui permet de gérer encore plus précisément la gestion d'erreur. 
+
+### Relancer une exception
+
+Parfois il est utile de tracer/logguer ce qui s'est mal déroulé.
```
Comment (rmasclef) [inline _posts/fr/2018-02-11-throwable-error-exception.md]: loguer ? logger ?

File: _posts/fr/2018-02-11-throwable-error-exception.md [hunk 11]
```diff
@@ -0,0 +1,374 @@
...
+    
+    public function create() 
+    {
+        try {
+            $user = new User();
+            $password = $this->passwordGenerator->generatePassword();
+            $user->setPassword($password);
+            
+            return $user;
+        } catch (Exception $exception) {
```
Comment (rmasclef) [inline _posts/fr/2018-02-11-throwable-error-exception.md]: \Exception

File: _posts/fr/2018-02-11-throwable-error-exception.md [hunk 12]
```diff
@@ -0,0 +1,374 @@
...
+            $this->logger->error('Une erreur est survenue pendant la creation d'un utilisateur. Raison: ' . $exception->getMessage());
```
Comment (rmasclef) [inline _posts/fr/2018-02-11-throwable-error-exception.md]: `$this->logger->error('Une erreur est survenue pendant la creation d\'un utilisateur. Raison: {exceptionMessage}.', ['exceptionMessage' => $exception->getMessage()]);`

File: _posts/fr/2018-02-11-throwable-error-exception.md [hunk 13]
```diff
@@ -0,0 +1,374 @@
...
+
+### Encapsuler un exception
+
+Il existe également l'encapsulation d'une `Exception` dans une autre `Exception` afin de créer un stack trace complète.
+
+```php
+try {
+    // mise à jour d'un contenu
+}
+catch (Exception $exception) {
```
Comment (rmasclef) [inline _posts/fr/2018-02-11-throwable-error-exception.md]: stay on above line

Comment (rmasclef) [reply _posts/fr/2018-02-11-throwable-error-exception.md]: \Exception

File: _posts/fr/2018-02-11-throwable-error-exception.md [hunk 14]
```diff
@@ -0,0 +1,374 @@
...
+    throw new UpdateContentException('Erreur de mise a jour du contenu.', 0, $exception);
```
Comment (rmasclef) [inline _posts/fr/2018-02-11-throwable-error-exception.md]: catch Exception -> throw new SpecializedException without knowing context ? :(

File: _posts/fr/2018-02-11-throwable-error-exception.md [hunk 15]
```diff
@@ -0,0 +1,374 @@
...
+> Dans cet exemple le script affichera le titre s'il est fourni
+> sinon il affichera le message d'erreur comme quoi il est obligatoire.
+
+Depuis PHP 7.1 il est également possible de spécifier plusieurs types d'`Exception` dans le catch en utilisant le caractère `|`
+
+```php
+try {
+    // Code
+}
+catch (OutOfBoundsException | LogicException $e) {
```
Comment (rmasclef) [inline _posts/fr/2018-02-11-throwable-error-exception.md]: Maybe you can talk about the priority ?

Explain (for example the importance of catching \Exception at the end ...) :

```
try {
   // Some code...
} catch (ExceptionType1 $e) {
   // Code to handle the exception
} catch (ExceptionType2 $e) {
   // Same code to handle the exception
} catch (Exception $e) {
   // ...
}
```

see https://wiki.php.net/rfc/multiple-catch

File: _posts/fr/2018-02-11-throwable-error-exception.md [hunk 16]
```diff
@@ -0,0 +1,374 @@
...
+class UserFactoryException extends RuntimeException {}
+
+interface PasswordGeneratorInterface
+{
+    public function generatePassword();
+}
+```
+
+> On peut voir ici que peu importe l'`Exception` qui se produit dans `$this->passwordGenerator->generatePassword()`
+> l'`Exception` qui sera remontée est une `UserFactoryException` qui nous informe que la création a échoué.
```
Comment (rmasclef) [inline _posts/fr/2018-02-11-throwable-error-exception.md]: échouée

File: _posts/fr/2018-02-11-throwable-error-exception.md [hunk 17]
```diff
@@ -0,0 +1,374 @@
...
+
+Pour attraper et gérer l'exception, il faut utiliser la structure `try` `catch`.
+```php
+try {
+    if (!$_GET['titre']) {
+        throw new Exception('Impossible d'afficher le titre. Le titre est requis.);
+    }
+    echo $_GET['titre'];
+}
+catch (Exception $e) {
```
Comment (rmasclef) [inline _posts/fr/2018-02-11-throwable-error-exception.md]: \Exception

Comment (instabledesign) [reply _posts/fr/2018-02-11-throwable-error-exception.md]: le `\` n'est pas obligatoire car il depend du namespace courant.
ici nous sommes déjà dans le namespace global

File: _posts/fr/2018-02-11-throwable-error-exception.md (hunk 15 above)
Comment (instabledesign) [reply _posts/fr/2018-02-11-throwable-error-exception.md]: yeah i forgot to talk about multiple catch but it's a different feature then `wildcard catch`
i'm gonna to add a paragraph to talk about multi catch thx :)

File: _posts/fr/2018-02-11-throwable-error-exception.md (hunk 11 above)
Comment (instabledesign) [reply _posts/fr/2018-02-11-throwable-error-exception.md]: we are again in global namespace

File: _posts/fr/2018-02-11-throwable-error-exception.md (hunk 12 above)
Comment (instabledesign) [reply _posts/fr/2018-02-11-throwable-error-exception.md]: my goal is not to explain the monolog usage in this article
i'll take the `\'` correction

**Comment by ElevenWilson** [comment]
28 words might be misspelled, please check them: amoutte, débogage, encapsuler, Encapsuler, eval, generatePassword, getMyObject, getPrevious, lang, l'encapsulation, LogicException, loguer, MyClass, MyObject, MyObjectException, ParseError, parsez, passwordGenerator, personalisée, rethrow, RuntimeException, throwable, Throwable, TypeError, typehint, UpdateContentException, url, UserFactoryException

File: _posts/fr/2018-02-11-throwable-error-exception.md (hunk 12 above)
Comment (rmasclef) [reply _posts/fr/2018-02-11-throwable-error-exception.md]: This is not monolog linked, its PSR 🍞

**Comment by ElevenWilson** [comment]
28 words might be misspelled, please check them: amoutte, débogage, encapsuler, Encapsuler, eval, generatePassword, getMyObject, getPrevious, lang, l'encapsulation, LogicException, loguer, MyClass, MyObject, MyObjectException, ParseError, parsez, passwordGenerator, personalisée, rethrow, RuntimeException, throwable, Throwable, TypeError, typehint, UpdateContentException, url, UserFactoryException

**Pull request merged**